### PR TITLE
Move the implementation of register_fake onto torch.library.Library

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124299
* #124200
* #124199
* #124134
* #124180
* #124092
* #124089
* #124071
* #124066
* __->__ #124065
* #124064
* #123937

Motivations:
- This makes things more consistent: using a Library object, you should
  be able to do all of the registration APIs that tie registrations to
  the lifetime of the Library.
- I need this for the next PR up in the stack, where we will have
  torch.library.register_fake support both CustomOpDef (from the new
  custom ops API) and other custom ops.

Test Plan:
- existing tests

### CODE REVIEW

File: torch/library.py [hunk 1]
```diff
@@ -125,6 +125,30 @@ def define(self, schema, alias_analysis="", *, tags=()):
...
+    def _register_fake(self, op_name, fn, _stacklevel=1):
+        r'''Registers the fake impl for an operator defined in the library.'''
+        source = torch._library.utils.get_source(_stacklevel + 1)
+        frame = sys._getframe(_stacklevel)
+        caller_module = inspect.getmodule(frame)
+        # Can be none if you call register_fake from somewhere there isn't a module
+        # (e.g. __main__)
+        caller_module_name = None if caller_module is None else caller_module.__name__
+
+        # TODO(rzou): We're gonna need to stage this change with torchvision,
```
Comment (albanD) [inline torch/library.py:137]: Not sure what this means?
Is it just: "TODO: update torchvision and remove this code that is here just for BC"

File: torch/library.py [hunk 2]
```diff
@@ -491,34 +515,23 @@ def register_fake(qualname, func=None, /, *, lib=None, _stacklevel=1):
...
-    if caller_module_name is not None and caller_module_name.startswith("torchvision."):
-        caller_module_name = None
-
-    def inner(func):
-        entry = torch._library.simple_registry.singleton.find(qualname)
-        if caller_module_name is not None:
-            func_to_register = _check_pystubs_once(func, qualname, caller_module_name)
-        else:
-            func_to_register = func
+    stacklevel = _stacklevel
```
Comment (albanD) [inline torch/library.py:518]: Why this rename here and not in the arg?
Private args in python are a lie anyways, everyone can see them haha

**Review by albanD — APPROVED** [review]
Bit more tricky than I was expecting, but sounds fair.

File: torch/library.py (hunk 1 above)
Comment (zou3519) [reply torch/library.py:137]: Yes. I didn't add the comment; it was moved from its original location

File: torch/library.py (hunk 2 above)
Comment (zou3519) [reply torch/library.py:518]: I needed to update the stacklevel below and didn't want to modify the `_stacklevel` argument

**Pull request closed without merging**